daisy/snapshot.go

```diff
@@ -0,0 +1,99 @@
+//  Copyright 2020 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package daisy
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"regexp"
+
+	daisyCompute "github.com/GoogleCloudPlatform/compute-image-tools/daisy/compute"
+	"google.golang.org/api/compute/v1"
+)
+
+var (
+	snapshotURLRgx = regexp.MustCompile(fmt.Sprintf(`^(projects/(?P<project>%[1]s)/)?global/snapshots/(?P<snapshot>%[2]s)$`, projectRgxStr, rfc1035))
+)
+
+// snapshotExists should only be used during validation for existing GCE snapshots
+// and should not be relied or populated for daisy created resources.
+func (w *Workflow) snapshotExists(project, snapshot string) (bool, DError) {
+	return w.snapshotCache.resourceExists(func(project string, opts ...daisyCompute.ListCallOption) (interface{}, error) {
+		return w.ComputeClient.ListSnapshots(project)
+	}, project, snapshot)
+}
+
+// Snapshot is used to create a GCE disk in a project.
+type Snapshot struct {
+	compute.Snapshot
+	Resource
+
+	sourceDiskProject string
+	sourceDiskZone    string
+	sourceDiskName    string
+}
+
+// MarshalJSON is a hacky workaround to prevent Snapshot from using compute.Snapshot's implementation.
+func (ss *Snapshot) MarshalJSON() ([]byte, error) {
+	return json.Marshal(*ss)
+}
+
+func (ss *Snapshot) populate(ctx context.Context, s *Step) DError {
+	var errs DError
+	ss.Name, errs = ss.Resource.populateWithGlobal(ctx, s, ss.Name)
+
+	ss.Description = strOr(ss.Description, fmt.Sprintf("Snapshot created by Daisy in workflow %q on behalf of %s.", s.w.Name, s.w.username))
+
+	if diskURLRgx.MatchString(ss.SourceDisk) {
+		ss.SourceDisk = extendPartialURL(ss.SourceDisk, ss.Project)
+	}
```

Comment:
Can you add some comments here? Here are some questions I think of when I see input validation that doesn't have an accompanying failure block: What happens when this doesn't match? Is it an expected case? A bug in our code? Bad input from the user?

Reply:
added